### PR TITLE
Feat/filter buttons

### PR DESCRIPTION
**Pull request by jcee-io**
So pretty large change here, I would like some input. Here are the things I've changed:

**General**
- Interface, MapView, and FilterButton now use PropTypes

**Data**
- Because we lack a solid database in development, the best way to handle filter buttons is to create fake data. I randomly assigned wifi, coffee, and open flags to all parklets in order to showcase this functionality. Let me know what you think of the structure of the data. since it simply added three properties that would determine the flags and how the filters would operate on state changes.

**SCSS/CSS**
- No hover effects on buttons, clicking it makes the animation a little funky
- Let me know what we should do with compiled style.css, does it belong there? 

**Interface**
- Interface has a handleFilters prop due to housing the FilterButtons
- Address is not a FilterButton, so a refactor in both Interface and FilterButton is implemented to establish address as a regular button
 
**FilterButton**
- toggleFilter function established
** For some reason <i> acts as a button when clicked so I added a recursive portion to resolve that, check it out
- address removed because it isn't a FilterButton

**App and MapView**
- Fetching parklet data now done in App (AKA App is now stateful). This is in order to correspond with the Interface component when using the filter buttons as a click handler
- Parklet state now separated from "parklet" to "activeParklet" and "hiddenParklet" to correspond with filters
** Mapview uses error and parklet state as props now due to App
** FilterButton now uses a prop that will call on the handler when clicked

P.S. Totally did finishing touches while tired, let me know if I can fix anything. Also apologies for the camelcasing of branches, it was already there and I didn't want to create a new branch.

P.P.S Getting rekt by tests, I'll fix my code tomorrow since the failed tests were mostly on the error PropType and other miscellaneous things. Also since filtering is completely random, I wasn't so sure how to test that.

### CODE REVIEW

File: client/src/App.jsx [hunk 1]
```diff
@@ -1,13 +1,83 @@
...
+  }
+
+  componentDidMount() {
+    // IMPORTANT: Until db is implemented, fake data will be used to generate flags
+    // for a parklet that is open, has coffee, or wifi by RNG, this will change for
+    // every refresh of the page
+    fetchParklets(PARKLETS_ENDPOINT)
+      .then((data) => {
+        // *******FAKE DATA GENERATOR************
+        const random = () => Math.random() > 0.5;
```
Comment (andino93) [inline client/src/App.jsx]: data generation should be a helper function. add it in a lib dir and import it since it can abstracted away and its not really part of the functionality of our react component.

File: client/src/App.jsx [hunk 2]
```diff
@@ -1,13 +1,83 @@
...
+        // remove above function once graphQL db is configured
+
+        this.setState({ activeParklets: data });
+      })
+      .then(() => this.handleFilters())
+      .catch(err => this.setState({ error: err }));
+  }
+
+  handleFilters() {
+    const open = document.querySelector('.open');
```
Comment (andino93) [inline client/src/App.jsx]: use react selectors instead of native JS on the dom

File: client/src/App.jsx [hunk 3]
```diff
@@ -1,13 +1,83 @@
...
+    const food = document.querySelector('.food');
```
Comment (andino93) [inline client/src/App.jsx]: use react selectors instead of native JS on the dom

File: client/src/App.jsx [hunk 4]
```diff
@@ -1,13 +1,83 @@
...
+    const wifi = document.querySelector('.wifi');
```
Comment (andino93) [inline client/src/App.jsx]: use react selectors instead of native JS on the dom

File: client/src/App.jsx [hunk 5]
```diff
@@ -1,13 +1,83 @@
...
+    const { activeParklets, hiddenParklets } = this.state;
+    const parklets = [...activeParklets, ...hiddenParklets];
+    const show = [];
+    const hide = [];
+
+    parklets.forEach((parklet) => {
```
Comment (andino93) [inline client/src/App.jsx]: make helper method

File: client/src/FilterButton.jsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    classList.add('btn-dark');
+  }
+
+  if (handler) {
+    handler();
+  }
+};
+
+const FilterButton = props => (
+  <button onClick={event => toggleFilter(event, props.handler)} className={`btn btn-dark ${props.type}`}>
```
Comment (andino93) [inline client/src/FilterButton.jsx]: this is wider than 100 chars

File: client/src/mapView.jsx [hunk 7]
```diff
@@ -1,42 +1,32 @@
...
     this.setState({ displayModal: !this.state.displayModal });
     return ('Pass me back into the modal component', clickedMarkerData);
   }
 
   render() {
-    const {
-      center, zoom, parklets, error, displayModal,
-    } = this.state;
+    const { center, zoom, displayModal } = this.state;
+    const { parklets, error } = this.props;
```
Comment (andino93) [inline client/src/mapView.jsx:28]: parklets and error are defined but not used

File: client/src/mapView.jsx [hunk 8]
```diff
@@ -1,42 +1,32 @@
...
   toggleModal(clickedMarkerData) {
     this.setState({ displayModal: !this.state.displayModal });
     return ('Pass me back into the modal component', clickedMarkerData);
   }
 
   render() {
-    const {
-      center, zoom, parklets, error, displayModal,
-    } = this.state;
+    const { center, zoom, displayModal } = this.state;
```
Comment (andino93) [inline client/src/mapView.jsx:27]: displayModal is defined but not used

**Review by andino93 — CHANGES_REQUESTED** [review]
Some changes requested but overall, solid. Sidenote, do you have a working linter? A lot of these errors should have been picked up by the linter.

**Comment by andino93** [comment]
Also @jcruzz, you're failing CI tests due to your proptype definitions. You also don't have any tests written for this PR.

File: client/src/App.jsx [hunk 9]
```diff
@@ -1,13 +1,83 @@
...
+      if (!parklet.open && open.classList.contains('btn-dark-outline')) {
```
Comment (andino93) [inline client/src/App.jsx]: This whole function is working directly with the DOM. Please restructure to the React way using the virtual DOM.

**Review by andino93 — CHANGES_REQUESTED** [review]

File: client/src/mapView.jsx (hunk 8 above)
Comment (mikehern) [reply client/src/mapView.jsx:27]: It's used on line 39:
` <ReactModal isOpen={displayModal} />`

File: client/src/mapView.jsx (hunk 7 above)
Comment (mikehern) [reply client/src/mapView.jsx:28]: This is on lines 34 and 37:
`{error &&
     <h3>Issue loading parklets: {error}</h3>
  }
     <MarkerCollection parklets={parklets} displayModal={this.toggleModal} />
`

File: client/src/App.jsx (hunk 2 above)
Comment (mikehern) [reply client/src/App.jsx]: This is diving a little bit into implementation detail - what kinds of data filtering methods do you both have in mind?

Off the top of my head, I was thinking the data would be structured so that each parklet contains those 3 properties.  And then filters would take the state, use a filter function, and on click, set the state.

Since all of the properties would exist on parklets, you could use higher-order functions to just filter and map the state.  Then you wouldn't need to split the state into filtered vs unfiltered.  Instead, state could just be comprised of a set of stackable active filters.

This is just a suggestion -- I also don't think we should manipulate the DOM directly with filters.

File: client/style.css [hunk 10]
```diff
@@ -51,19 +50,15 @@
   border-style: solid;
   border-width: 2px;
   border-color: #10b535;
-  -webkit-transition-duration: 300ms;
```
Comment (mikehern) [inline client/style.css:54]: Just curious, why did you remove the browser polyfills in the last commit?

**Review by mikehern — CHANGES_REQUESTED** [review]
More implementation detail than code style, I don't think we should be manipulating the DOM directly.

I made a suggestion related to how to architect `state` by keeping it pure through higher-order functions, but I'd like to hear your thoughts on what you both think of this.

File: client/src/mapView.jsx (hunk 7 above)
Comment (andino93) [reply client/src/mapView.jsx:28]: Oops, I didn't see them since it wasn't part of the changed files. My bad.

File: client/src/mapView.jsx (hunk 8 above)
Comment (andino93) [reply client/src/mapView.jsx:27]: Oops, I didn't see them since it wasn't part of the changed files. My bad.

Comment (jcee-io) [reply client/src/mapView.jsx:27]: No worries, didn't touch any of that since filter buttons were my main concern haha.

File: client/src/App.jsx (hunk 2 above)
Comment (jcee-io) [reply client/src/App.jsx]: It isn't a direct manipulation since it's a state change that would result in calling render() again, but I see where you're coming from. This approach uses the existence of a class, that is if the FilterButtons are .btn-dark-outline or not to attempt to hide the parklet or send it to the else case, where we'll show it. The selectors are merely just to find out the event.

That said, I can always use states to see whether or not open, food, and wifi filters are active. This would seem cleaner. Thanks for the suggestion!

File: client/style.css (hunk 10 above)
Comment (jcee-io) [reply client/style.css:54]: This was a side effect of compiling, I'm not so sure why it did that, but I merely just removed the hover selector.

File: client/src/App.jsx (hunk 2 above)
Comment (andino93) [reply client/src/App.jsx]: It's a conflict because it's not separating concerns. In react, you can still access classNames via the event object if you really need them. It's better to use props though because it has much clearer data flow rather than reading the Dom from react.

Comment (andino93) [reply client/src/App.jsx]: Also, while you're still using a state method to set the state, you're directly manulating the Dom by using `document.querySelector`

**Comment by jcee-io** [comment]
NEW CHANGES
**Fake Data Generator**
- Now is its own function located in project root as fakeFilterFlags.js

**Parklet Refresh On Filter Change**
- Helper function to sort it as active or hidden put in its own function
- Instead of using DOM selectors, states are now used
** States are based on the type of filter button used (the type prop is passed into handleFilters)

TODO:
- Pass current tests
- Write more tests
- Discuss possible improvements to this feature

**Comment by andino93** [comment]
@mikehern @jcruzz I'm going to close this PR pending TODO list completion. That cool with you guys?

File: fakeFilterFlags.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+module.exports = (data) => {
```
Comment (andino93) [inline fakeFilterFlags.js]: Please use ES2015+ import/export statements per style guide

**Review by andino93 — CHANGES_REQUESTED** [review]
Please also organize `fakeFilterFlags.js` in a utils dir to keep things organized instead of just dumping in the root of the project.

**Comment by andino93** [comment]
Closing this until agreed changes are made so CI stops running for each push.
